### PR TITLE
Revert "Revert "Add max voting power to sei-cosmos (#103)" (#112)"

### PR DESCRIPTION
**Pull request by udpatil**
This reverts commit 4bc3ca67bba0b054a4719374532baa71f9459e93.
## Describe your changes and provide context
This reintroduces the max voting power changes along with fixes for parallelization

### CODE REVIEW

File: x/staking/keeper/val_state_change.go [hunk 1]
```diff
@@ -67,8 +69,10 @@ func (k Keeper) BlockValidatorUpdates(ctx sdk.Context) []abci.ValidatorUpdate {
 		if err != nil {
 			panic(err)
 		}
-		delegatorAddress := sdk.MustAccAddressFromBech32(dvvTriplet.DelegatorAddress)
-
+		delegatorAddress, err := sdk.AccAddressFromBech32(dvvTriplet.DelegatorAddress)
```
Comment (BrandonWeng) [inline x/staking/keeper/val_state_change.go:72]: isn't this the same as before?
```
// MustAccAddressFromBech32 calls AccAddressFromBech32 and panics on error.
func MustAccAddressFromBech32(address string) AccAddress {
	addr, err := AccAddressFromBech32(address)
	if err != nil {
		panic(err)
	}

	return addr
}
```

**Review by BrandonWeng — APPROVED** [review]

File: x/staking/keeper/val_state_change.go (hunk 1 above)
Comment (udpatil) [reply x/staking/keeper/val_state_change.go:72]: hm, yeah not sure why this change was made, I'll just change that back lol

**Comment by yzang2019** [comment]
@udpatil This change actually breaks the upgrade since it's missing a migration path after we change the proto and add new params, it cause the upgrade to stuck and throw errors:

> panic: UnmarshalJSON cannot decode empty bytes [recovered]
> 	panic: UnmarshalJSON cannot decode empty bytes
> 
> goroutine 816 [running]:
> go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.func1()
> 	/root/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.9.0/trace/span.go:363 +0x2a
> go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc0022d0d80, {0x0, 0x0, 0x382ca40?})
> 	/root/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.9.0/trace/span.go:402 +0x8dd
> panic({0x1c38520, 0xc00416fcd0})
> 	/usr/lib/go-1.18/src/runtime/panic.go:838 +0x207
> github.com/cosmos/cosmos-sdk/x/params/types.Subspace.Get({{0x28a15e0, 0xc000484670}, 0xc0005983a8, {0x287d7f8, 0xc0002a6550}, {0x287d848, 0xc0002a6690}, {0xc0005aa540, 0x7, 0x29}, ...}, ...)
> 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/params/types/subspace.go:109 +0x307
> github.com/cosmos/cosmos-sdk/x/staking/keeper.Keeper.MaxVotingPowerRatio(...)
> 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/keeper/params.go:44
> github.com/cosmos/cosmos-sdk/x/staking/keeper.Keeper.GetParams({{0x287d7f8, 0xc0002a64e0}, {0x28a15e0, 0xc000484670}, {0x2897fc0, 0xc00056e360}, {0x7f04d45eab40, 0xc000573a20}, {0x28a1418, 0xc0001300d8}, ...}, ...)
> 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/keeper/params.go:77 +0x6a8
> github.com/cosmos/cosmos-sdk/x/staking/keeper.Keeper.ApplyAndReturnValidatorSetUpdates({{0x287d7f8, 0xc0002a64e0}, {0x28a15e0, 0xc000484670}, {0x2897fc0, 0xc00056e360}, {0x7f04d45eab40, 0xc000573a20}, {0x28a1418, 0xc0001300d8}, ...}, ...)
> 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/keeper/val_state_change.go:109 +0xbb
> github.com/cosmos/cosmos-sdk/x/staking/keeper.Keeper.BlockValidatorUpdates({{0x287d7f8, 0xc0002a64e0}, {0x28a15e0, 0xc000484670}, {0x2897fc0, 0xc00056e360}, {0x7f04d45eab40, 0xc000573a20}, {0x28a1418, 0xc0001300d8}, ...}, ...)
> 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/keeper/val_state_change.go:27 +0x7d
> github.com/cosmos/cosmos-sdk/x/staking.EndBlocker({{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, {0x1a9c0c07, ...}, ...}, ...}, ...)
> 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/abci.go:25 +0xff
> github.com/cosmos/cosmos-sdk/x/staking.AppModule.EndBlock(...)
> 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/module.go:174
> github.com/cosmos/cosmos-sdk/types/module.(*Manager).EndBlock(_, {{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, ...}, ...}, ...)
> 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/types/module/module.go:543 +0x482
> github.com/sei-protocol/sei-chain/app.(*App).EndBlocker(_, {{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, ...}, ...}, ...)
> 	/sei-protocol/sei-chain/app/app.go:916 +0x78
> github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).EndBlock(_, {{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, ...}, ...}, ...)
> 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/baseapp/abci.go:186 +0x11d
> github.com/sei-protocol/sei-chain/app.(*App).EndBlock(_, {{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, ...}, ...}, ...)
> 	/sei-protocol/sei-chain/app/abci.go:32 +0x145
> github.com/sei-protocol/sei-chain/app.(*App).ProcessBlock(_, {{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, ...}, ...}, ...)
> 	/sei-protocol/sei-chain/app/app.go:1281 +0xa25
> github.com/sei-protocol/sei-chain/app.(*App).ProcessProposalHandler.func1()
> 	/sei-protocol/sei-chain/app/app.go:947 +0xc5
> created by github.com/sei-protocol/sei-chain/app.(*App).ProcessProposalHandler
> 	/sei-protocol/sei-chain/app/app.go:946 +0x1f8
>

**Comment by udpatil** [comment]
> @udpatil This change actually breaks the upgrade since it's missing a migration path after we change the proto and add new params, it cause the upgrade to stuck and throw errors:
> 
> > panic: UnmarshalJSON cannot decode empty bytes [recovered]
> > 	panic: UnmarshalJSON cannot decode empty bytes
> > 
> > goroutine 816 [running]:
> > go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.func1()
> > 	/root/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.9.0/trace/span.go:363 +0x2a
> > go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc0022d0d80, {0x0, 0x0, 0x382ca40?})
> > 	/root/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.9.0/trace/span.go:402 +0x8dd
> > panic({0x1c38520, 0xc00416fcd0})
> > 	/usr/lib/go-1.18/src/runtime/panic.go:838 +0x207
> > github.com/cosmos/cosmos-sdk/x/params/types.Subspace.Get({{0x28a15e0, 0xc000484670}, 0xc0005983a8, {0x287d7f8, 0xc0002a6550}, {0x287d848, 0xc0002a6690}, {0xc0005aa540, 0x7, 0x29}, ...}, ...)
> > 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/params/types/subspace.go:109 +0x307
> > github.com/cosmos/cosmos-sdk/x/staking/keeper.Keeper.MaxVotingPowerRatio(...)
> > 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/keeper/params.go:44
> > github.com/cosmos/cosmos-sdk/x/staking/keeper.Keeper.GetParams({{0x287d7f8, 0xc0002a64e0}, {0x28a15e0, 0xc000484670}, {0x2897fc0, 0xc00056e360}, {0x7f04d45eab40, 0xc000573a20}, {0x28a1418, 0xc0001300d8}, ...}, ...)
> > 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/keeper/params.go:77 +0x6a8
> > github.com/cosmos/cosmos-sdk/x/staking/keeper.Keeper.ApplyAndReturnValidatorSetUpdates({{0x287d7f8, 0xc0002a64e0}, {0x28a15e0, 0xc000484670}, {0x2897fc0, 0xc00056e360}, {0x7f04d45eab40, 0xc000573a20}, {0x28a1418, 0xc0001300d8}, ...}, ...)
> > 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/keeper/val_state_change.go:109 +0xbb
> > github.com/cosmos/cosmos-sdk/x/staking/keeper.Keeper.BlockValidatorUpdates({{0x287d7f8, 0xc0002a64e0}, {0x28a15e0, 0xc000484670}, {0x2897fc0, 0xc00056e360}, {0x7f04d45eab40, 0xc000573a20}, {0x28a1418, 0xc0001300d8}, ...}, ...)
> > 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/keeper/val_state_change.go:27 +0x7d
> > github.com/cosmos/cosmos-sdk/x/staking.EndBlocker({{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, {0x1a9c0c07, ...}, ...}, ...}, ...)
> > 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/abci.go:25 +0xff
> > github.com/cosmos/cosmos-sdk/x/staking.AppModule.EndBlock(...)
> > 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/x/staking/module.go:174
> > github.com/cosmos/cosmos-sdk/types/module.(*Manager).EndBlock(_, {{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, ...}, ...}, ...)
> > 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/types/module/module.go:543 +0x482
> > github.com/sei-protocol/sei-chain/app.(*App).EndBlocker(_, {{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, ...}, ...}, ...)
> > 	/sei-protocol/sei-chain/app/app.go:916 +0x78
> > github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).EndBlock(_, {{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, ...}, ...}, ...)
> > 	/root/go/pkg/mod/github.com/sei-protocol/sei-cosmos@v0.1.382/baseapp/abci.go:186 +0x11d
> > github.com/sei-protocol/sei-chain/app.(*App).EndBlock(_, {{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, ...}, ...}, ...)
> > 	/sei-protocol/sei-chain/app/abci.go:32 +0x145
> > github.com/sei-protocol/sei-chain/app.(*App).ProcessBlock(_, {{0x2894a98, 0xc0044c0810}, {0x28a3c28, 0xc00445d2c0}, {{0x0, 0x0}, {0x7ffe8936a901, 0x3}, 0x4b0, ...}, ...}, ...)
> > 	/sei-protocol/sei-chain/app/app.go:1281 +0xa25
> > github.com/sei-protocol/sei-chain/app.(*App).ProcessProposalHandler.func1()
> > 	/sei-protocol/sei-chain/app/app.go:947 +0xc5
> > created by github.com/sei-protocol/sei-chain/app.(*App).ProcessProposalHandler
> > 	/sei-protocol/sei-chain/app/app.go:946 +0x1f8
> > 

Good callout, can you put a revert PR if it's blocking rn and I can approve (afk rn)

**Comment by yzang2019** [comment]
This is not urgent, it's just some interesting findings when I'm testing the upgrade, we can create a ticket and follow up next week

**Comment by LCyson** [comment]
> This is not urgent, it's just some interesting findings when I'm testing the upgrade, we can create a ticket and follow up next week

sg thanks, can assign that ticket to me, will add a migration handler when i get a chance

**Pull request merged**